### ISSUE TITLE
empty menus

### ISSUE THREAD

**Issue by hg42** [ssborbis/ContextSearch-web-ext#151]
For some time now, my toolbar menu is empty for about 50% of invocations.
In these cases I see the search box, some empty space (about two lines) and the settings button.
Usually it happens a few times in sequence before it changes.
It happens with or without text in the search field.
It happens with or without animations (but it looks similar).
In many cases I see that the menu is long and then immediately collapses (very short, kind of flashing). Not sure if there are search engines or if the lines are empty.

Something similar happens, when pressing the sidebar button, but only from time to time. This menu is usually completely empty (I think).
Just now I pressed my hotkey for the sidebar a couple of times and I now see only two lines of engines followed by the settings button while the menu window has it's usual size.
If I close it the button is kind of dead (cannot open again), while the other menus still work. But I can use the hotkey to open it and still get the two lines.
At the same time other pages work.

seems to work for Chrome

FF67.0a1

**Comment by hg42** [comment]
I also disabled my CSS rules with no effect

**Comment by ssborbis** [comment]
I've seen this with the toolbar menu. I tried nailing down the problem, but it was so irregular I couldn't catch any repeatable error. Seemed to start around FF64 or 65

**Comment by ssborbis** [comment]
I reworked the toolbar code a bit. I haven't been able to trigger the bug with the new code, but I've never seen it happen on the sidebar. If it's still happening after the latest push, post / post a pic of your sidebar settings and I'll do more testing.

**Comment by hg42** [comment]
I can reproduce the sidebar "completely empty" part... (but not the "only two lines" part)
If I open the sidebar on page A (I tried youtube, amazon, aliexpress), then open the options from there and then select the tab of A again, the sidebar is empty.
This is still happening after the last pull.

The toolbar menu behaves similar to yesterday but the issue happens much less frequently.

I just tested with columns (yesterday only single columns).
That's interesting...when I chose 3 columns I sometimes got an empty menu in the usual width (like my 8 column menu, but I think it's the default width?) and sometimes it's the same width like the correct menu. The "default" width seems to be much more frequent. For the single column menu I always get the same width (it's about one icon wider than the 8 columns).

From the behaviour of the toolbar menu and the sidebar menu I would conclude that these are two separate issues, though both might be based on the same cause.

**Comment by ssborbis** [comment]
> If I open the sidebar on page A (I tried youtube, amazon, aliexpress), then open the options from there and then select the tab of A again, the sidebar is empty.

Ok, that was an easy fix. I think there was some glitch with the toolbar window in an earlier FF version when opening options so I set the body visibility to none first. I never changed it after implementing the sidebar.

> The toolbar menu behaves similar to yesterday but the issue happens much less frequently.
> ... when I chose 3 columns I sometimes got an empty menu in the usual width

I've only been testing with 3 columns ( to force the width / resize code ) and I haven't had a single bug since updating the code. I wonder if it's a difference of hardware causing some timing issues, or a config difference. I'll keep on it.

**Comment by ssborbis** [comment]
Trying some new methods. The toolbar is a major pain to troubleshoot because it closes as soon as the window loses focus making debugging damn-near impossible. 

I also noticed the sitesearch wasn't showing the full width of text for url paths. Dunno when that started because I rarely use it, but that's fixed along with some sidebar height weirdness when opening a less-populated folder from a larger folder using the floating overlay setting.

**Comment by hg42** [comment]
see this video:
[2019-02-22-toolbar-menu-empty.zip](https://github.com/ssborbis/ContextSearch-web-ext/files/2892194/2019-02-22-toolbar-menu-empty.zip)

the newest code changed something, last time the clipped yellow folder wasn't there, only a gray background.
So it apparently the contents is now drawn but the menu size is the minimum.

The sidebar is ok now, thanks for the fix.

**Comment by ssborbis** [comment]
I'm still trying a few things, but I did release 1.9.6 as a bit of a hotfix because of previous version failing to install on FF66+. I made a minor change to the toolbar code since your last test so maybe that fixed the issue. If not, I'll try to get something ready for 1.9.7

**Comment by hg42** [comment]
I just upgraded to 1.9.6.1 (06d2f67a932407bbe6931c674de085d0f5f63c9c), the issue still exists for the toolbar.

It seems to be a problem with the event sequence in some way:

Sometimes I see a big menu (not sure if already filled) that collapses instantly to the empty version (gray with Search box and Settings button).
Sometimes I see a small menu (I _think_ I saw it also with the button) that expands to the big.
After some invocations it's without collapsing or expansion. I guess it's because the processing is faster with caches filled, code optimized (JIT compilation, branch-optimization) etc.

If I get the small version and enter something in the search box, the suggestions appear first and after a fraction of a second the Rest of the menu is also filled. It's just slow enough to see how it's filled and growing from top to bottom.

**Comment by ssborbis** [comment]
I've got a new method to size the toolbar to test, but I'm also in the middle of moving the sidebar to use the dock/undock code so it's a bit sketchy right now. I'll make the push, but I did have a bug when first trying to move / open the sidebar from the tab. The way the sidebar expands from the tab is also a bit different, based on the quadrant it's in.

**Comment by ssborbis** [comment]
@hg42 

Any improvement in the lastest patches? I haven't seen an empty menu yet. YMMV

**Comment by githubeing** [comment]
tl;dr
the bug still reproduces on my side. sometimes the toolbar opens empty, i got it on camera, see the [screencast-2019-03-19_22.44.15.zip](https://github.com/ssborbis/ContextSearch-web-ext/files/2985343/screencast-2019-03-19_22.44.15.zip)
environment is Chromium Version 73.0.3683.75 (Official Build) Built on Ubuntu , running on Ubuntu 18.10 (64-bit)

**Comment by ssborbis** [comment]
@githubeing 

Are you using the latest push or the previous release (1.9.6) ?

**Comment by githubeing** [comment]
Version 1.9.6.1

**Comment by hg42** [comment]
@ssborbis sorry for answering late, I had to concentrate on some things

> Any improvement in the lastest patches? I haven't seen an empty menu yet. YMMV

I just pulled 1.9.7b 014eceaef7004de08c10a18550f21926d9cb162d

The sidebar seems to work properly.
First I thought it fails, because of the small menu, but I simply had to resize the menu once (I had forgotten, that the resize function was in the queue lately and didn't see the lower right corner).

On the other hand, I still get the empty menu in the toolbar. I think this behavior hasn't changed for quite a lot of commits.

But there is some progress.
For some time now, I sensed a dependency on the way I click the button. But I couldn't find a rule, yet.

Because of this I made a little test. If clicking slowly, meaning pressing the button for some longer time ( I guess about 0.5 sec or more) there is never an empty menu.
I think, I usually click very short.
I guess, in those cases, when I get an empty menu, the mouse button is already released before the menu is open.
Visually, fast clicking doesn't always produce an empty menu. But I guess, it's not really dependent on the visual part, but more on the number of entries already filled.

**Comment by hg42** [comment]
I also tested if it depends on the single column, but it does not.

I now have 95 active entries in the list (including folder entries).
The behavior might depend on the number of entries. E.g. the processing might not have finished when the mouse button changes it's state.

**Comment by ssborbis** [comment]
I just cannot get the toolbar to glitch with the newer code. I've tried on FF65, FF68 nightly on windows and FF65 on linux. I've opened and closed the menu hundreds of times and haven't seen a single failure. I'll keep looking.

**Comment by hg42** [comment]
I didn't see an empty menu for some time now...works nice...
I think we can close this...
thanks

**Comment by githubeing** [comment]
i just got this bug reproduced on v1.11 for ubuntu,chromium
the first time i clicked the icon the toolbar menu was empty, no search engines were there. then i re-clicked it again and the second time everything was ok, the search engines were there.
btw, (i'm not sure but it seems to me that) i've got this bug reproduced once not long ago, maybe a couple of weeks ago.

**Comment by githubeing** [comment]
@ssborbis the bug still reproduces for me, ext v.1.18, ubuntu 19.10 64-bit, chromium Version 79.0.3945.79 (Official Build) snap (64-bit).
can you reproduce that?

**Comment by ssborbis** [comment]
@githubeing 

I can reproduce, but usually only on the first toolbar menu opening of the session. Not sure if it's a delay in loading a resource or something else. This bug has lingered. I'll work on it some more before the next release.
